### PR TITLE
NEXT-12688 - Versioning entities

### PR DESCRIPTION
**Pull request by Dominik28111**

### CODE REVIEW

File: guides/plugins/plugins/framework/data-handling/versioning-entities.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+## Prerequisites
+
+In order to add your own versioned entities for your plugin, you first need a plugin as base. Therefore, you can refer to the [Plugin Base Guide](../../plugin-base-guide.md).
+
+Furthermore you should have a look at our [Adding custom complex data](./add-custom-complex-data.md) guide, since this guide is built upon it.
+
+## Adjust migration
+
+First of all, we have to a new columns to our table, `version_id` which has to be a primary key beside the `id`.
```
Comment (lernhart) [inline guides/plugins/plugins/framework/data-handling/versioning-entities.md:16]:
```suggestion
First of all, we have to add a new column to our table: `version_id` which in union with the `id` field replaces the primary key.
```

File: guides/plugins/plugins/framework/data-handling/versioning-entities.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        ...
+    ]);
+}
+```
+{% endcode %}
+
+## Create and merge version
+
+In this section we will create a new version of our entity which will create a new entry in the database with our updated values.
+When we merge a particular version, all versions before the merged version are deleted. The version can
```
Comment (lernhart) [inline guides/plugins/plugins/framework/data-handling/versioning-entities.md:47]: The version can ... ??

File: guides/plugins/plugins/framework/data-handling/versioning-entities.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+```
+{% endcode %}
+
+## Create and merge version
+
+In this section we will create a new version of our entity which will create a new entry in the database with our updated values.
+When we merge a particular version, all versions before the merged version are deleted. The version can 
+In the example below, we are using a service where we injected a `swag_example.repository`.
+
+*Note: If an entity is not loaded in the specified version, the live version is reverted to.*
```
Comment (lernhart) [inline guides/plugins/plugins/framework/data-handling/versioning-entities.md:50]: [...] is reverted to what ? Or is it [...] reverted, too ?

File: guides/plugins/plugins/framework/data-handling/versioning-entities.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+{% code title="<plugin root>/src/" %}
+```php
+public function exampleVersioning(Context $context)
```
Comment (lernhart) [inline guides/plugins/plugins/framework/data-handling/versioning-entities.md:54]:
```suggestion
public function exampleVersioning(Context $context): void
```

File: guides/plugins/plugins/framework/data-handling/versioning-entities.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+As you can see above, we first created a new `ExampleEntity` with the description 'This is an example'.
+
+Then we created a new version of our entity with the appropriate repository method `createVersion` and as arguments the id of our created entity and the context.
+This method returns the id of our new entity version, which we have stored in a variable.
+
+Next, we used the `createWithVersionId` method of our `Context` to create a new context with our new versionId assigned to it.
+This new `Context` is used to update the `ExampleEntity`. In our case we have updated the description to 'This is our new description'.
+By using the updated context with the new `versionId`, the DAL knows that we want to update this version of our entity.
+
+Now we did repository searches with our versioned `Context` and our normal.
```
Comment (lernhart) [inline guides/plugins/plugins/framework/data-handling/versioning-entities.md:102]:
```suggestion
Subsequently, we searched the repository with the original context and our new versioned context.
```

File: guides/plugins/plugins/framework/data-handling/versioning-entities.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+In the first search result with our normal `Context` we get the first version of our entity that we created at the beginning.
```
Comment (lernhart) [inline guides/plugins/plugins/framework/data-handling/versioning-entities.md:103]:
```suggestion
In the first search result, using the original context, we get the first version of our entity, which we created at the beginning.
```

File: guides/plugins/plugins/framework/data-handling/versioning-entities.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+The second search result with the updated `Context` w get the updated entity.
```
Comment (lernhart) [inline guides/plugins/plugins/framework/data-handling/versioning-entities.md:104]:
```suggestion
With the second search result we get the updated entity, using our new versioned context.
```

File: guides/plugins/plugins/framework/data-handling/versioning-entities.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+Lastly, we used the repository method `merge` with our versionId, which deletes all versions before this one. 
+The merged version is now our new live version. From now on we can find it with the normal `Context`.
```
Comment (lernhart) [inline guides/plugins/plugins/framework/data-handling/versioning-entities.md:107]:
```suggestion
The merged version is now our new live version. From now on we can find it without using a versioned context.
```

File: guides/plugins/plugins/framework/data-handling/versioning-entities.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+## Versioning with foreign keys
+
+If you have an entity with foreign keys, your foreign keys also need the version field.
+In this example we're using an inherited field. If you are not familiar with this, head over to our [Field inheritance](./field-inheritance.md) guide.
```
Comment (lernhart) [inline guides/plugins/plugins/framework/data-handling/versioning-entities.md:112]:
```suggestion
If you have an entity with foreign keys, your foreign keys also need to be versioned.
In this example we're using an inherited field. If you are not familiar with inheritance, head over to our [Field inheritance](./field-inheritance.md) guide.
```

**Review by lernhart — CHANGES_REQUESTED** [review]

File: guides/plugins/plugins/framework/data-handling/versioning-entities.md [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+Then we created a new version of our entity with the appropriate repository method `createVersion` and as arguments the id of our created entity and the context.
+This method returns the id of our new entity version, which we have stored in a variable.
+
+Next, we used the `createWithVersionId` method of our `Context` to create a new context with our new versionId assigned to it.
+This new `Context` is used to update the `ExampleEntity`. In our case we have updated the description to 'This is our new description'.
+By using the updated context with the new `versionId`, the DAL knows that we want to update this version of our entity.
+
+Subsequently, we searched the repository with the original context and our new versioned context.
+In the first search result, using the original context, we get the first version of our entity, which we created at the beginning.
+With the second search result we get with the updated entity, using our new versioned context.
```
Comment (lernhart) [inline guides/plugins/plugins/framework/data-handling/versioning-entities.md:102]:
```suggestion
With the second search result we get the updated entity, using our new versioned context.
```

**Review by lernhart — CHANGES_REQUESTED** [review]

**Comment by JoshuaBehrens** [comment]
Why is `parent_version_id` NULL in SQL but with a Required-flag in DAL?

File: guides/plugins/plugins/framework/data-handling/versioning-entities.md [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+        ...
+    ]);
+}
+```
+{% endcode %}
+
+## Next steps
+
+The next thing you can do is to do something with inherited fields, which have already been mentioned in this article.
+For this head over to our [Field inheritance](./field-inheritance.md) guide.
```
Comment (PaddyS) [inline guides/plugins/plugins/framework/data-handling/versioning-entities.md:149]: Remove

**Review by PaddyS — CHANGES_REQUESTED** [review]

**Pull request merged**